### ISSUE TITLE
Exporting more of the types

### ISSUE THREAD

**Issue by brecert** [heyheyhello/haptic#11]
It'd be nice if all of the exposed types were exported in some way.
I often use types directly from libraries when writing functions and types and haptic does not currently export all of it's types.

Thanks to typescript, types typically aren't opaque so it's not essential to export every type that's exposed through functions and other objects, but it would be nice if they were, 

I'm assuming they aren't exported because they aren't meant to be public or visible, or possibly that they are redundant types, and I think even if it was hidden from documentation and wasn't supported by the expected guarantees of semantic versioning it'd still be nice to have.

While specifying the types for each of the `when` status view functions here would be better, this is kind of the example usage I'm thinking about in this issue.
```tsx
// These could be imported from `haptic/stdlib,`
// but they aren't exported so they need to be copied here for full compatibility
type Component = (...args: unknown[]) => El;
type El = Element | Node | DocumentFragment;

type AsyncStatus = "unresolved" | "resolved" | "rejected";
function Async<T>(promise: Promise<T>, views: Record<AsyncStatus, Component>) {
  const status = signal.anon<AsyncStatus>("unresolved");
  const data = signal.anon<T | null>(null);

  promise
    .then((res) => {
      status("resolved");
      data(res);
    })
    .catch((err) => {
      status("rejected");
      data(err);
    });

  return when(
    wire(($) => status($)),
    views
  );
}
```

**Comment by tsugabloom** [comment]
I love when packages export their types and completely agree. I think I didn't export these specific types because they're...not good types. I had a _very bad time_ trying to type SInuous' `h()` call which uses multiple overloaded function declarations that take _some_ assortment of elements, nodes, fragments, components, etc.

I feel like I settled on https://github.com/heyheyhello/haptic/blob/b1b21b76e355630341ace8a6c5da0fb33efb5d00/src/dom/h.ts#L5-L7

As kind of a shrug between what the TypeScript compiler wants and what is meaningful to us as developers/people. For instance, `El` could (maybe should?) be `HTMLElement | SVGElement | Node | ...` and other people will want to simply use `Node` for _everything_ because it's the base class. This is what a lot of Sinuous' internal APIs do.

The issue with TypeScript is... complicated. I went and dug up issues from 2020 I had trying to wrap the overloaded h call: https://github.com/microsoft/TypeScript/issues/38698.


In Sinuous there were types that tried to account for everything:

- https://github.com/luwes/sinuous/blob/ece2c65c19939ec5742f999cfba6c83295c58118/packages/sinuous/shared.d.ts#L3-L18
- https://github.com/luwes/sinuous/blob/master/packages/sinuous/h/src/index.d.ts#L7

Now today I wonder if `unknown` would be just as meaningful...

Sorry this got long - I agree with you and I'll find a way to export `h()`'s types. I just don't know if the `El` type is right.

Here's how I'd type your example:

```ts
type AsyncStatus = 'unresolved' | 'resolved' | 'rejected';
type WhenViews<T extends string> = { [k in T]?: () => Node }

function Async<T>(promise: Promise<T>, views: WhenViews<AsyncStatus>) {
  const status = signal.anon<AsyncStatus>('unresolved');
  const data = signal.anon<T | null>(null);
  ...
}
```

Which could even be said to be more accurate since functions in `when()`'s view object take no parameters `() => ...` while "Components" are `(...args: unknown[]) => ...`.

I'll export types when I'm doing dev tomorrow. We can always evolve them as needed :)
